wasm/src/validator.rs

```diff
@@ -0,0 +1,400 @@
+                 otherwise panic message would be less informative"
+            )
+        };
+        ($l:literal as $t:ty) => {
+            $l.parse::<$t>().dbg_expect(concat!(
+                "Failed to parse `",
+                $l,
+                "` as `",
+                stringify!($t),
+                "`"
+            ))
+        };
+    }
+
+    /// Macro to create [`Grant`](crate::data_model::prelude::Grant) and
+    /// [`Revoke`](crate::data_model::prelude::Revoke) instructions validation
+    /// for a given token type using its [`Validate`](super::traits::Validate) implementation.
+    ///
+    /// Generated code will do early return with `instruction` validation verdict
+    /// only if it is a `Grant` or `Revoke` instruction.
+    ///
+    /// Otherwise it will proceed with the rest of the function.
+    ///
+    /// # Syntax
+    ///
+    /// ```no_run
+    /// validate_grant_revoke!(<Token1, Token2, ...>, (authority_ident, instruction_ident));
+    /// ```
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// #[derive(Token, Validate)]
+    /// #[validate(Creator)]
+    /// struct CanMintAssetsWithDefinition {
+    ///     asset_definition_id: <AssetDefinition as Identifiable>::Id,
+    /// }
+    ///
+    /// #[entrypoint(params = "[authority, instruction]")]
+    /// pub fn validate(authority: <Account as Identifiable>::Id, instruction: Instruction) -> Verdict {
+    ///    validate_grant_revoke!(<CanMintAssetsWithDefinition>, (authority, instruction));
+    ///    // ...
+    /// }
+    /// ```
+    #[macro_export]
+    macro_rules! validate_grant_revoke {
+        (< $($token:ty),+ $(,)?>, ($authority:ident, $instruction:ident $(,)?)) => {
+            if let $crate::data_model::prelude::Instruction::Grant(grant) = &$instruction {
+                if let $crate::data_model::prelude::Value::PermissionToken(permission_token) =
+                    <
+                        $crate::data_model::prelude::EvaluatesTo<$crate::data_model::prelude::Value>
+                        as
+                        $crate::EvaluateOnHost
+                    >::evaluate_on_host(&grant.object)
+                    .dbg_expect("Failed to evaluate `Grant` object")
+                {$(
+                    if let Ok(concrete_token) =
+                        <$token as ::core::convert::TryFrom<_>>::try_from(
+                            <
+                                $crate::data_model::permission::token::Token as ::core::clone::Clone
+                            >::clone(&permission_token)
+                        )
+                    {
+                        return <$token as ::iroha_wasm::validator::traits::Validate>::validate_grant(
+                            &concrete_token,
+                            &$authority
+                        );
+                    }
+                )+}
+                if let $crate::data_model::prelude::Instruction::Revoke(revoke) = &$instruction {
```

Comment:
at first I thought this is not properly indented but no this is useless. please use match statement. There is no test checking that `revoke` works? Let's add a test, this is scary

Reply:
it should look like this (you can format to your liking):
```rs

macro_rules! validate_grant_revoke {
    (< $($token:ty),+ $(,)?>, ($authority:ident, $instruction:ident $(,)?)) => {
        match &$instruction {
            $crate::data_model::prelude::Instruction::Grant(grant) => {
                let value = <$crate::data_model::prelude::EvaluatesTo<$crate::data_model::prelude::Value> as $crate::EvaluateOnHost>::evaluate_on_host(&grant.object).dbg_expect("Failed to evaluate `Grant` object");

                if let $crate::data_model::prelude::Value::PermissionToken(permission_token) = value {$(
                    if let Ok(concrete_token) = <$token as ::core::convert::TryFrom<_>>::try_from(permission_token) {
                        return <$token as ::iroha_wasm::validator::traits::Validate>::validate_grant(&concrete_token, &$authority);
                    }
                )+}
            }
            $crate::data_model::prelude::Instruction::Revoke(revoke) => {
                let value = <$crate::data_model::prelude::EvaluatesTo<$crate::data_model::prelude::Value> as $crate::EvaluateOnHost>::evaluate_on_host(&revoke.object).dbg_expect("Failed to evaluate `Revoke` object");

                if let $crate::data_model::prelude::Value::PermissionToken(permission_token) = value {$(
                    if let Ok(concrete_token) = <$token as ::core::convert::TryFrom<_>>::try_from(permission_token) {
                        return <$token as ::iroha_wasm::validator::traits::Validate>::validate_revoke(&concrete_token, &$authority);
                    }
                )+}
            }
        }
    };
}
```

cloning was redundant as well, no?

Reply:
Clonning is redundant only on last step. In general case there can be a lot of token types which we should check. For that case we need to clone `permission_token` to try it on the next `try_from()`

Reply:
>  There is no test checking that revoke works? Let's add a test, this is scary

Created #3256.

Also previously we had pretty no `Revoke` checks. So on current `iroha2-dev` it's possible to `Revoke` any token from anyone